### ISSUE TITLE
Using marshmallow and automatically generate definitions

### ISSUE THREAD

**Issue by kaboo-dev** [rochacbruno/flasgger#296]
**TL;DR**: 

Is it possible to generate a spec definition automatically from a marshmallow schema and add it to my spec defined when I do `swag = Flasgger(app)` ?

--------

Currently, i'm writting a Flask app, using : 

- Flask-SQLAlchemy
- Flask-Marshmallow
- Marshmallow-SQLAlchemy
- Flasgger

I have built the application the following way : 

inside a `server.py` file:
```python
from flask import Flask
from flasgger import Swagger
from flask_sqlalchemy import SQLAlchemy
from flask_marshmallow import Marshmallow
# [...]
app = Flask(__name__)
swag = Swagger(app)
db = SQLAlchemy(app)
ma = Marshmallow(app)
# [...]
from submodule.routes import *
```
and the routes (defined with the `@app.route` decorator, inside a submodule). 

These routes extend the "db.Model" SQLAlchemy class, and in order to add definitions to my spec, I'm currently using the @Flasgger::definition decorator (and writting a YAML file in the docstring of the decorated class, re-declaring all of the properties). This means that I'm rewritting the information that already exists within the `marshmallow.ModelShema` object.

I'd like to know if it is possible to generate the same spec definitions, automatically from the marshmallow schemas, and add them to the spec.

(NB: The "template from apidocs" pattern described in [here](https://github.com/rochacbruno/flasgger/blob/master/examples/apispec_example.py) can't be used in my case as I define the Flasgger object (`swag = Flasgger(app)`)  before my routes (and my schemas) are imported.  

--------

Ideally, it would be nice to have a decorator that decorates Marshmallow classes, and adds them to a spec as definitions:

something like 
```python
@swag.definition_from_marshmallow(definition_name='User', endpoint='myendpoint')
class UserSchema(ma.ModelSchema):
    class Meta:
         model = User
```

that would result in a User definition, that can be then referenced.

![Screenshot from 2019-04-15 18-53-54](https://user-images.githubusercontent.com/17289748/56150786-eeb0f200-5faf-11e9-806e-59834caa070b.png)

**Comment by gbrener** [comment]
If all you're interested in is referencing the marshmallow definitions from within a `SwaggerView` method's docstring (as I am), this seems to do the trick:

```
from flasgger import Schema, SwaggerView

class UserSchema(Schema):
    ... (fields go here) ...

class UserView(SwaggerView):
    methods = ['GET']
    definitions = {'UserSchema': UserSchema}

    def get(self):
        """Get users.
        ---
        responses:
          200:
            description:
            schema:
              $ref: '#/definitions/UserSchema'
        """
        ... (rest of get method's code goes here) ...

    ... (rest of UserView class's code goes here) ...
```

You might be able to use this approach to get what you want. For example, you could write a `SwaggerView` class that does nothing but reference all of your marshmallow schemas (you'd need to reference each schema in a docstring, however).

**Comment by mbierma** [comment]
Based on #341 and starting with release 0.9.4.dev2 (or later), you can add your marshmallow schema using the `@swag_from` decorator:

```
from flasgger import Schema
from http import HTTPStatus

class TestSchema(Schema):
    number = fields.Int()

@app.route('/test', methods=['POST'])
@swag_from({
    'responses': {
        HTTPStatus.OK.value: {
            'schema': TestSchema
        }
    }
}, validation=True)
def test():
    """
    Test POST 
    ---
    """
    data = request.json
    # TODO
```

**Comment by gskolysz** [comment]
#341 feature works for OAS v2. For version 3+, the schema is not being rendered :(.

**Comment by Jonyorker** [comment]
I can confirm on my codebase that using 

```
@swag_from({
    'responses': {
        HTTPStatus.OK.value: {
            'schema': TestSchema
        }
    }
})
```

Does not render the schema if OAS3 is enabled.

**Comment by mirecl** [comment]
Hi, I am create new PR https://github.com/flasgger/flasgger/pull/396 for easy usage Marshmallow Schema in flasgger
